### PR TITLE
NET-1420: Encapsulate validation logic

### PR DESCRIPTION
**Pull request by seanterry**
https://seaseducation.atlassian.net/browse/NET-1420
---
### Problem
As a developer, I want to use FluentValidation to validate objects of any type without needing to re-implement any FluentValidation plumbing for common scenarios so that I can accomplish most validation with a single dependency.

### Solution
- Added an adapter to encapsulate boilerplate validation logic.
- Allowed for multiple validators to be defined for a type.
- Added a marker interface that indicates a type must have at least one validator defined that will throw an exception at runtime if the developer forgot to add one.

### Additional Notes
Assumes validators will be supplied via dependency injection container.

### CODE REVIEW

File: src/Validation/IValidationAdapter.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+        async Task ValidateAndThrow( T instance, CancellationToken cancellationToken )
+        {
+            var result = await Validate( instance, cancellationToken );
+
+            if ( !result.IsValid )
+            {
+                throw new ValidationException( result.Errors );
+            }
+        }
```
Comment (seanterry) [inline src/Validation/IValidationAdapter.cs:44]: Point of interest: this is a relatively new C# language feature that allows interfaces to define a default implementation.

Comment (peterbillings7) [reply src/Validation/IValidationAdapter.cs:44]: Does a class that implements the interface have to specify that it is using the default implementation from the interface?  Also, if you didn't want to use the default, would you have to override it as with a virtual or abstract method?

File: src/Validation/ValidationAdapter.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    /// <typeparam name="T">Type to validate.</typeparam>
+
+    public class ValidationAdapter<T> : IValidationAdapter<T>
+    {
+        private readonly IEnumerable<IValidator<T>> validators;
+
+        /// <summary>
+        /// Constructs and adapter for encapsulating validation functionality.
+        /// </summary>
+        /// <param name="validators">Collection of validators defiend for the type.</param>
```
Comment (peterbillings7) [inline src/Validation/ValidationAdapter.cs:28]: there is a typo in the word 'defined' in the validators param description

File: src/Validation/IValidationAdapter.cs (hunk 1 above)
Comment (seanterry) [reply src/Validation/IValidationAdapter.cs:44]: > Does a class that implements the interface have to specify that it is using the default implementation from the interface?

No. From the developer's perspective, implementing the method becomes completely optional. You'll notice on the `ValidationAdapter<T>` class that it doesn't contain a defition for the `ValidateAndThrow` method.

> Also, if you didn't want to use the default, would you have to override it as with a virtual or abstract method?

You would only need to implement the method in the class to not use the default. Under the covers, when the compiler sees a default implementation, it checks the class to see if it has the method defined. If it doesn't find it, it adds it.

**Review by peterbillings7 — APPROVED** [review]
I just noticed a typo, which I made a comment about.  Otherwise, everything looks good.

File: src/Validation/IValidationAdapter.cs (hunk 1 above)
Comment (peterbillings7) [reply src/Validation/IValidationAdapter.cs:44]: I see now.  That is pretty cool, and likely extremely useful.

**Pull request merged**